rust/src/records.rs

```diff
@@ -0,0 +1,267 @@
+}
+
+impl<'a> IntoStatic for Channel<'a> {
+    type Static = Channel<'static>;
+
+    fn into_static(self) -> Self::Static {
+        Channel {
+            id: self.id,
+            schema_id: self.schema_id,
+            topic: self.topic.into_static(),
+            message_encoding: self.message_encoding.into_static(),
+            metadata: self
+                .metadata
+                .iter()
+                .map(|(k, v)| (k.clone().into_static(), v.clone().into_static()))
+                .collect(),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Message<'a> {
+    pub channel_id: u16,
+    pub sequence: u32,
+    pub log_time: Timestamp,
+    pub publish_time: Timestamp,
+    pub data: Cow<'a, [u8]>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Chunk<'a> {
+    pub message_start_time: Timestamp,
+    pub message_end_time: Timestamp,
+    pub uncompressed_size: u64,
+    pub uncompressed_crc: u32,
+    pub compression: Cow<'a, str>,
+    pub records: Cow<'a, [u8]>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct MessageIndex {
+    pub channel_id: u16,
+    pub records: Vec<(Timestamp, u64)>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct ChunkIndex<'a> {
+    pub message_start_time: Timestamp,
+    pub message_end_time: Timestamp,
+    pub chunk_start_offset: u64,
+    pub chunk_length: u64,
+    pub message_index_offsets: BTreeMap<u16, u64>,
+    pub message_index_length: u64,
+    pub compression: Cow<'a, str>,
+    pub compressed_size: u64,
+    pub uncompressed_size: u64,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Attachment<'a> {
+    pub log_time: Timestamp,
+    pub create_time: Timestamp,
+    pub name: Cow<'a, str>,
+    pub content_type: Cow<'a, str>,
+    pub data: Cow<'a, [u8]>,
+    #[serde(skip_serializing)]
+    pub crc: u32,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct AttachmentIndex<'a> {
+    pub offset: u64,
+    pub length: u64,
+    pub log_time: Timestamp,
+    pub create_time: Timestamp,
+    pub data_size: u64,
+    pub name: Cow<'a, str>,
+    pub content_type: Cow<'a, str>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Statistics {
+    pub message_count: u64,
+    pub schema_count: u16,
+    pub channel_count: u32,
+    pub attachment_count: u32,
+    pub metadata_count: u32,
+    pub chunk_count: u32,
+    pub message_start_time: Timestamp,
+    pub message_end_time: Timestamp,
+    pub channel_message_counts: BTreeMap<u16, u64>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone)]
+pub struct Metadata<'a> {
+    pub name: Cow<'a, str>,
+    pub metadata: BTreeMap<Cow<'a, str>, Cow<'a, str>>,
+}
+
+impl<'a> IntoStatic for Metadata<'a> {
```

Comment:
What the heck is IntoStatic and how is it possible to make something static? Is this like constexpr?

Reply:
A third-party trait: https://docs.rs/lifetime/latest/lifetime/trait.IntoStatic.html , used to simplify creating a "owned" form of all of our record structs.

It is possible to make some struct `<'a>` into `<'static>` if you can ensure that none of the fields are borrows that use that lifetime parameter. For example for a `Cow<'a, str>`, IntoStatic would be implemented something like:

```rust
  fn into_static(self) -> Cow<'static, str> {
      Cow::Owned(self.to_string())
  }
```